File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.py
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        graph, ans = defaultdict(set), []
        inDegrees = [0] * len(recipes)
        
        for i in range(len(ingredients)):
            for gr in ingredients[i]:
                graph[gr].add(i)
                inDegrees[i] += 1
                
        q = deque([sup for sup in supplies])
        
        while q:
            p = q.popleft()
            sup = recipes[p] if str(p).isdecimal() else p
            if sup in recipes:
                ans.append(sup)
            for ne in graph[sup]:
                inDegrees[ne] -= 1
                if inDegrees[ne] == 0:
                    q.append(ne)
                    
        return ans
        

File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.cpp
class Solution {
public:
    vector<string> findAllRecipes(vector<string>& recipes, vector<vector<string>>& ingredients, vector<string>& supplies) {
        int n = recipes.size();
        unordered_map<string, vector<string>> adj;
        unordered_map<string, int> mp;
        vector<int>deg(n);
        for (int i = 0; i < n; i++) {
            mp[recipes[i]] = i;
            for (auto gd: ingredients[i]) {
                adj[gd].push_back(recipes[i]);
                deg[i]++;
            }
        }
        queue<string> q;
        for (int i = 0; i < n; i++)
            if (deg[i] == 0)
                q.push(recipes[i]);
        for (auto x: supplies)
            q.push(x);
        set<string> ans;
        while (!q.empty()) {
            string node = q.front();
            ans.insert(node);
            q.pop();
            for (auto u: adj[node]) {
                if (--deg[mp[u]] == 0)
                    q.push(u);
            }
        }
        vector<string> res;
        for (auto x: recipes)
            if (ans.count(x))
                res.push_back(x);
        return res;
    }
};